src/storage/mod.rs

```diff
@@ -20,6 +20,9 @@ pub mod storages;
 #[cfg(test)]
 mod tests;
 
+/// A trait that guarantees that mutable storage getter can be used on distinct indices from different threads.
```

Comment:
I think this needs a clearer warning in which cases one may implement it, but I don't see any reason we couldn't do that after merging.